File: Best_Research/#Done/Rust/LearningRust/Units/Units/src/units/time/display.rs
use std::fmt;

use crate::internal::format_unit_value;

use super::quantity::{Time, TimeUnit};

impl Time {
    #[must_use]
    pub const fn display_as(self, unit: TimeUnit) -> TimeDisplay {
        TimeDisplay {
            value: self,
            unit,
            precision: None,
        }
    }

    #[must_use]
    pub const fn display_as_precision(self, unit: TimeUnit, precision: usize) -> TimeDisplay {
        TimeDisplay {
            value: self,
            unit,
            precision: Some(precision),
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct TimeDisplay {
    value: Time,
    unit: TimeUnit,
    precision: Option<usize>,
}

impl fmt::Display for TimeDisplay {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = match self.unit {
            TimeUnit::Seconds => self.value.as_seconds(),
            TimeUnit::Milliseconds => self.value.as_milliseconds(),
            TimeUnit::Microseconds => self.value.as_microseconds(),
            TimeUnit::Nanoseconds => self.value.as_nanoseconds(),
            TimeUnit::Minutes => self.value.as_minutes(),
            TimeUnit::Hours => self.value.as_hours(),
        };

        format_unit_value(formatter, value, self.unit.symbol(), self.precision)
    }
}

impl fmt::Display for Time {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.display_as(TimeUnit::Seconds).fmt(formatter)
    }
}

File: Best_Research/#Done/C++/Units-CppXmake/units/time/display.cpp
#include "units/display/display.hpp"

#include <ostream>

namespace units::time
{

    const char* symbol(TimeUnit unit)
    {
        switch (unit)
        {
        case TimeUnit::Seconds:
            return "s";
        case TimeUnit::Milliseconds:
            return "ms";
        case TimeUnit::Microseconds:
            return "us";
        case TimeUnit::Nanoseconds:
            return "ns";
        case TimeUnit::Minutes:
            return "min";
        case TimeUnit::Hours:
            return "h";
        }

        return "";
    }

    double valueInUnit(Time value, TimeUnit unit)
    {
        switch (unit)
        {
        case TimeUnit::Seconds:
            return value.asSeconds();
        case TimeUnit::Milliseconds:
            return value.asMilliseconds();
        case TimeUnit::Microseconds:
            return value.asMicroseconds();
        case TimeUnit::Nanoseconds:
            return value.asNanoseconds();
        case TimeUnit::Minutes:
            return value.asMinutes();
        case TimeUnit::Hours:
            return value.asHours();
        }

        return value.asSeconds();
    }

    std::ostream& operator<<(std::ostream& stream, Time value)
    {
        return stream << value.displayAs(TimeUnit::Seconds);
    }

} // namespace units::time
